### PR TITLE
[JENKINS-26100] Switch buildEnvVars and others to Run

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-26100](https://issues.jenkins-ci.org/browse/JENKINS-26100)

Downstream of https://github.com/jenkinsci/jenkins/pull/2730

cc @reviewbybees esp @MarkEWaite @jglick

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -24,7 +24,8 @@
   <inceptionYear>2007</inceptionYear>
 
   <properties>
-    <jenkins.version>1.625.3</jenkins.version>
+    <jenkins-core.version>2.58-20170501.171905-3</jenkins-core.version> <!-- TODO: Switch to release once https://github.com/jenkinsci/jenkins/pull/2730 is merged and released -->
+    <jenkins-war.version>2.58-20170501.171927-3</jenkins-war.version> <!-- TODO: Switch to release once https://github.com/jenkinsci/jenkins/pull/2730 is merged and released -->
     <java.level>7</java.level>
```
Comment (jglick) [inline pom.xml]: 8

File: src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java [hunk 2]
```diff
@@ -470,7 +470,7 @@ protected String getCacheEntry() {
     }
 
     protected static File getCacheDir(String cacheEntry) {
-        Jenkins jenkins = Jenkins.getInstance();
+        Jenkins jenkins = Jenkins.getInstanceOrNull();
```
Comment (jglick) [inline src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java]: Most of these should actually be `getInstance()` with no null check, but good enough for now.

Comment (abayer) [inline src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java]: I went for the most stupidly non-logic-altering Findbugs fix I could. =) Wasn't 100% sure there weren't some cases where they might be expecting the existing behavior with a possibly null `jenkins` instance, so hey, better safe than sorry.

**Comment by jglick** [comment]
```
Hudson Parameterized Trigger plugin v2.4 failed to load.
 - subversion v1.7 is missing. To fix, install v1.7 or later.
```

Not your fault, just a sign that this plugin’s `Jenkinsfile` is missing the usual idiom to verify that its tests run against Jenkins 2.12+, e.g.:

```groovy
buildPlugin(jenkinsVersions: [null, '2.32.3'])
```

**Comment by MarkEWaite** [comment]
Last time I enabled the plugin compat tests, it failed because the combination of tests and compilation took more than an hour to complete.  The ci.jenkins.io jobs (correctly) limit themselves to not more than an hour elapsed time for the total job.

I'll enable plugin compat testing again (it is in the Jenkinsfile as a comment), and we'll see if it performs better now.

**Comment by jglick** [comment]
But `jenkinsVersions` will be run in parallel, and the timeout is wall clock time, so why would it matter?

**Comment by MarkEWaite** [comment]
With plugin compat enabled, it now does not timeout (maybe a false recollection on my part).  It reports findbugs warnings on the 2.46.1 build which it does not report on other builds.  Unfortunately, my first attempts to duplicate that result locally have failed.  My findbugs is silent, even when I set ```jenkins.version=2.46.1```.

I suspect I won't have time to explore this further until after the end of the working day today.  If someone else can educate me how to duplicate the problem from my command line compile (or IDE), I'm happy to investigate why findbugs warnings are detected for Jenkins 2.46.1 based builds, but not detected for the actual base Jenkins version.

I'm also fine if someone can educate me how to not fail on findbugs warnings for Jenkins 2.46.1, while continuing to fail on findbugs warnings on the actual base Jenkins version.  My maven skills are not as strong as I'd like yet.

**Comment by jglick** [comment]
The most common introduced FB error is due to @stephenc’s messing around with `Jenkins.getInstance`. Either switch to `getActiveInstance`, if appropriate, or leave the `getInstance` call with a null check and add

```java
@SuppressFBWarnings(value="RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE", justification="TODO 1.653+ switch to Jenkins.getInstanceOrNull")
```

Yes this is obnoxious which is why I asked him to leave existing methods alone, and rely on deprecations if necessary, but he insisted…

**Comment by jglick** [comment]
FTR my NetBeans action binding for checking FindBugs:

    mvn -DskipTests=false clean compile findbugs:check

**Comment by MarkEWaite** [comment]
I'm not sure that action binding will work with the git plugin's current pom.  As far as I can tell, once ```-DskipTests=false``` is passed as an argument, then findbugs:check is a no-op.  I think that is reasonable (if you're skipping tests, you probably don't want to run findbugs), but it surprised me in this case.

My solution was ```-Dtest=CGitTest findbugs:check findbugs:gui```.  Runs a fast test, and reports findbugs results.

**Comment by jglick** [comment]
[This](https://github.com/jenkinsci/plugin-pom/blob/2258cef69660ca6081400dc5827eb44673a2d3d9/pom.xml#L887-L898) works in other plugins. FindBugs is skipped only if `-DskipTests` or `-DskipTests=true` is used. Omitting, or passing `-DskipTests=false`, runs FindBugs.

**Comment by MarkEWaite** [comment]
Thanks for the education!  Inattentional blindness problem on my part.  I saw the ```-DskipTests= ``` in your action binding and didn't look at the ```false``` value.  Nice technique.

File: pom.xml [hunk 3]
```diff
@@ -24,8 +24,9 @@
   <inceptionYear>2007</inceptionYear>
 
   <properties>
-    <jenkins.version>1.625.3</jenkins.version>
-    <java.level>7</java.level>
+    <jenkins-core.version>2.58-20170502.192524-8</jenkins-core.version> <!-- TODO: Switch to release once https://github.com/jenkinsci/jenkins/pull/2730 is merged and released -->
+    <jenkins-war.version>2.58-20170502.192544-8</jenkins-war.version> <!-- TODO: Switch to release once https://github.com/jenkinsci/jenkins/pull/2730 is merged and released -->
```
Comment (jglick) [inline pom.xml]: 2.61

Comment (jglick) [inline pom.xml]: 2.60 I meant.

File: src/main/java/hudson/plugins/git/GitChangeSet.java [hunk 4]
```diff
@@ -420,13 +421,16 @@ private boolean hasHudsonTasksMailer() {
     }
 
     private boolean isCreateAccountBasedOnEmail() {
-        Hudson hudson = Hudson.getInstance();
-        if (hudson == null) {
+        Jenkins jenkins = Jenkins.getInstance();
+        if (jenkins == null) {
```
Comment (jglick) [inline src/main/java/hudson/plugins/git/GitChangeSet.java]: Maybe just revert changes in this file?

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 5]
```diff
@@ -1255,8 +1255,12 @@ private void computeChangeLog(GitClient git, Revision revToBuild, TaskListener l
...
     }
 
-    public void buildEnvVars(AbstractBuild<?, ?> build, java.util.Map<String, String> env) {
-        super.buildEnvVars(build, env);
+    @Override
+    public void buildEnvVars(AbstractBuild<?, ?> build, Map<String, String> env) {
+        buildEnvironment(build, env);
+    }
+
+    public void buildEnvironment(Run<?, ?> build, java.util.Map<String, String> env) {
```
Comment (jglick) [inline src/main/java/hudson/plugins/git/GitSCM.java]: Put in the usual TODOs.

File: src/main/java/hudson/plugins/git/GitChangeSet.java (hunk 4 above)
Comment (abayer) [inline src/main/java/hudson/plugins/git/GitChangeSet.java]: 'k

File: src/main/java/hudson/plugins/git/GitSCM.java (hunk 5 above)
Comment (abayer) [inline src/main/java/hudson/plugins/git/GitSCM.java]: 'k

**Comment by Letty-Liang** [comment]
looking forward to be published soon

**Comment by abayer** [comment]
@reviewbybees done

@MarkEWaite Can I persuade you to merge this? =)

**Comment by MarkEWaite** [comment]
It will be several days (at least) before I have a chance to review it.  Can the merge wait until next week?  What if it needs to wait another week beyond that?

**Comment by abayer** [comment]
@MarkEWaite No rush at all - just wanted to make sure this wasn't lost.

File: pom.xml [hunk 6]
```diff
@@ -218,11 +218,15 @@
...
     <dependency><!-- we contribute AbstractBuildParameters for Git if it's available -->
-      <groupId>org.jvnet.hudson.plugins</groupId>
+      <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>parameterized-trigger</artifactId>
-      <version>2.4</version>
+      <version>2.33</version>
       <optional>true</optional>
+<<<<<<< HEAD
+   </dependency>
+=======
```
Comment (jglick) [inline pom.xml]: Bad merge conflict resolution!

**Comment by abayer** [comment]
@MarkEWaite Any news? =)

**Comment by MarkEWaite** [comment]
@abayer, sorry, but no progress on this pull request yet.  Is it urgent, or blocking other work that I should place it ahead of other things in my queue (like the submodule checkout regression or the credential exclusion issue)?

**Pull request merged**